rlmolecule/run.py

```diff
@@ -0,0 +1,137 @@
+            rollout_nodes=2, rollout_ntasks_per_node=6,
+            policy_nodes=1, policy_gpus=2,
+            **kwargs):
+        """
+        *hpc_env*: Computing environment for which a submission script will be 
+            generated and submitted. Allowed values: 'eagle'
+        *working_dir*: Directory where submission scripts and log files 
+            will be stored
+        *policy_env*: string to load the environment used for the GPU nodes
+        *rollout_env*: string to load the environment used for the CPU nodes
+        """
+        self.hpc_env = hpc_env
+        self.job_name = job_name
+        self.out_file = out_file
+        self.working_dir = working_dir
+        os.makedirs(self.working_dir, exist_ok=True)
+        self.time = time
+        self.account = account
+        self.start_policy_script = start_policy_script
+        self.start_rollout_script = start_rollout_script
+        self.policy_env = policy_env
+        self.rollout_env = rollout_env
+        self.rollout_nodes = rollout_nodes
+        self.rollout_ntasks_per_node = rollout_ntasks_per_node
+        self.policy_nodes = policy_nodes
+        self.policy_gpus = policy_gpus
+
+    def setup_hpc_submission_scripts(self, submit=False):
+        if self.hpc_env == "eagle":
+            submission_str = self.generate_eagle_submission_string()
+            command = f"sbatch {self.out_file}"
+        else:
+            print(f"ERROR: not yet setup to run on '{self.hpc_env}'")
+            print("Quitting")
+            sys.exit()
+
+        print(f"Writing submissing script to {self.out_file}")
+        with open(self.out_file, 'w') as out:
+            out.write(submission_str)
+
+        if submit:
+            print(f"Running {command}")
+            subprocess.check_call(command, shell=True)
+
+    def generate_eagle_submission_string(self):
+        submission_str = f"""#!/bin/bash
+#SBATCH --account={self.account}
+#SBATCH --time={self.time}
+#SBATCH --job-name={self.job_name}
+# --- Policy Trainer ---
+#SBATCH --nodes={self.policy_nodes}
+#SBATCH --gres=gpu:{self.policy_gpus}
+# --- MCTS Rollouts ---
+#SBATCH hetjob
+#SBATCH -N {self.rollout_nodes}
+
+export WORKING_DIR={self.working_dir}
+export START_POLICY_SCRIPT="$SLURM_SUBMIT_DIR/$JOB/.policy.sh"
+export START_ROLLOUT_SCRIPT="$SLURM_SUBMIT_DIR/$JOB/.rollout.sh"
+
+cat << "EOF" > "$START_POLICY_SCRIPT"
+#!/bin/bash
+{self.policy_env}
+python -u {self.start_policy_script}
+EOF
+
+cat << "EOF" > "$START_ROLLOUT_SCRIPT"
+#!/bin/bash
+{self.rollout_env}
+python -u {self.start_rollout_script}
+EOF
+
+chmod +x "$START_POLICY_SCRIPT" "$START_ROLLOUT_SCRIPT"
+
+srun --pack-group=0 \
+     --job-name="az-policy" \
+     --output=$WORKING_DIR/gpu.%j.out \
+     "$START_POLICY_SCRIPT" &
+
+srun --pack-group=1 \
+     --ntasks-per-node=6 \
+     --job-name="az-rollout" \
+     --output=$WORKING_DIR/mcts.%j.out \
+     "$START_ROLLOUT_SCRIPT"
+"""
+        return submission_str
```

Comment:
> This looks good, thanks for the work on getting the new code running on Eagle! Were you able to submit things and did they update on yuma?

Yep! They updated on yuma.

> I'm a bit concerned that the package-wide run / config scripts are too Eagle / HPC specific, and in general might be premature optimization that will make editing harder later on. For instance, the structure of the submission scripts will have to change when we use tf serving.

Yeah that makes sense. The main goals I was thinking with the config file setup is to 1) make sure none of the parameters are hard-coded in a python script so that the AZ or problem-specific parameters can be attached with a given run (and for reproducibility), and 2) try to reduce hard-coded setup code (e.g., connecting to the DB) and make it easier to share those configurations across scripts. 

I was envisioning having another section within the `run_config` section for different cases of running the code such as `tf_serving_settings` that would have the specific settings needed to make that happen. But yeah, likely premature and too specific to our architecture. 

> Would it be too much work to for now just have a minimal example of how to run the QED example on Eagle? I.e., it's OK if the submission script isn't fully resuable for other allocations or examples. Definitely turn up the max molecule size and try running this on ~10 nodes?

No problem, and yes I'll try that now.

Reply:
Way to jump into this code base, @jlaw9 !  Trial by fire is the best. : )

I agree with Peter at a high level, that having a barebones minimal, very Eagle-specific, example that works as an sbatch script (i.e., solves a large-ish molecule problem) might the right starting point.  My guess is that after using a minimal example a few times, we'll have a better idea of where to optimize / build interfaces versus just demonstrating how something might be done.  Does that make sense?  Happy to also discuss this on Teams or whatever if that's helpful.

Thanks for taking a stab at this!!

Reply:
I don't know if this is helpful context, but you might take a look at the RLLib library which we use a lot on other projects for scaling up RL, because they have this same issue of trying to getting their framework running on different architectures.

Slurm: https://docs.ray.io/en/master/cluster/slurm.html
Kubernetes:  https://docs.ray.io/en/master/cluster/kubernetes.html

You can see the slurm example is very lightweight, and totally independent of the Kubernetes stuff, and I think this is for good reason.  They did release a little python interface for Slurm specfically, but even that seems too general purpose to be useful (e.g., no support for heterogeneous job types).  

My hunch is that this might be a good design direction for us, too.

Reply:
> Way to jump into this code base, @jlaw9 ! Trial by fire is the best. : )

Thanks @davebiagioni, that's what I figured :) 

You probably saw, but I made a minimum working example script here: https://github.com/jlaw9/rlmolecule/tree/qed_yuma_mwe/examples/qed
I'll try to do the same thing for the gym examples next.

> I don't know if this is helpful context, but you might take a look at the RLLib library which we use a lot on other projects for scaling up RL, because they have this same issue of trying to getting their framework running on different architectures.

Yeah I thought there may be other people with the same challenge, I just wasn't sure where to look. Seems like a great library and could work well for our different applications. I'll have to take a closer look.